### PR TITLE
MGMT-8281: subsystem print resources when subsystem tests fail

### PR DESCRIPTION
**Pull request by nmagnezi**
# Assisted Pull Request

## Description

To make it easier (or actually, possible) to look into a failure from and
understand what happened in the backend, the subsystem tests should print
all resources when a test fails (clusters, hosts, infraenvs).

## List all the issues related to this PR

- [ ] New Feature
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Assignees

/cc @
/cc @

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] Reviewers have been listed
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- [ ] Are the title and description (in both PR and commit) meaningful and clear?
- [ ] Is there a bug required (and linked) for this change?
- [ ] Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

**Comment by nmagnezi** [comment]
/assign @osherdp 
/assign @filanov 

Find an example of how the output would look for failed tests ([intentionally](https://github.com/openshift/assisted-service/pull/2952)) here:
* [subsystem-aws](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_assisted-se[…]-assisted-service-master-subsystem-aws/1461433664465276928)
* [kubeapi-aws](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_assisted-se[…]d-service-master-subsystem-kubeapi-aws/1461433664511414272)

File: subsystem/kubeapi_test.go [hunk 1]
```diff
@@ -564,6 +564,62 @@ func randomNameSuffix() string {
...
+			aciList               hiveext.AgentClusterInstallList
+			agentList             v1beta1.AgentList
+			infraEnvList          v1beta1.InfraEnvList
+			bareMetalHostList     bmhv1alpha1.BareMetalHostList
+			nmStateConfigList     v1beta1.NMStateConfigList
+			clusterImageSetList   hivev1.ClusterImageSetList
+			clusterDeploymentList hivev1.ClusterDeploymentList
+		)
+		err = client.List(ctx, &agentList, k8sclient.InNamespace(Options.Namespace))
+		Expect(err).To(BeNil())
```
Comment (filanov) [inline subsystem/kubeapi_test.go:580]: Suggesting not to fail in this function, after all you want to collect everything you can here. 
maybe use multi error to collect all the errors and print them in the end.

File: subsystem/kubeapi_test.go [hunk 2]
```diff
@@ -690,6 +746,7 @@ var _ = Describe("[kube-api]cluster installation", func() {
 	})
 
 	AfterEach(func() {
+		printCRs(ctx, kubeClient)
```
Comment (filanov) [inline subsystem/kubeapi_test.go:749]: instead of using it everywhere maybe we can use Init or some other method to define it once?

**Review by osherdp — CHANGES_REQUESTED** [review]

File: subsystem/kubeapi_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (filanov) [reply subsystem/kubeapi_test.go:580]: no difference, just looks different, i think that `To(BeNil())` looks cleaner.

[the remark this answers is not in the document]
Comment (osherdp) [reply subsystem/kubeapi_test.go:580]: so let's just pick one of the options within perimeter of 3 lines? :smile:

File: subsystem/kubeapi_test.go [hunk 3]
```diff
@@ -564,6 +564,62 @@ func randomNameSuffix() string {
...
+			bareMetalHostList     bmhv1alpha1.BareMetalHostList
+			nmStateConfigList     v1beta1.NMStateConfigList
+			clusterImageSetList   hivev1.ClusterImageSetList
+			clusterDeploymentList hivev1.ClusterDeploymentList
+		)
+		err = client.List(ctx, &agentList, k8sclient.InNamespace(Options.Namespace))
+		Expect(err).To(BeNil())
+		aList, err1 := json.Marshal(agentList)
+		Expect(err1).ShouldNot(HaveOccurred())
+		By(fmt.Sprintf("failed test '%s' created the agents: %s", GinkgoT().Name(), aList))
```
[the remark this answers is not in the document]
Comment (nmagnezi) [reply subsystem/kubeapi_test.go:583]: I understand, but do we have another way to include the printed output for created resources?

File: subsystem/kubeapi_test.go (hunk 1 above)
Comment (nmagnezi) [reply subsystem/kubeapi_test.go]: Done

File: subsystem/kubeapi_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (osherdp) [reply subsystem/kubeapi_test.go]: Logging doesn't work?
Or printing if it's not possible

[the remark this answers is not in the document]
Comment (filanov) [reply subsystem/kubeapi_test.go]: can try to use print as well

File: subsystem/kubeapi_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nmagnezi) [reply subsystem/kubeapi_test.go]: Done

File: subsystem/kubeapi_test.go (hunk 2 above)
Comment (nmagnezi) [reply subsystem/kubeapi_test.go:767]: It is currently used just a couple of times, one per each `AfterEach()`, same as DB cleanup, etc. 
the `"[kube-api]cluster installation"` and `"bmac reconcile flow"`test suits are currently indepdandent

File: subsystem/kubeapi_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (nmagnezi) [reply subsystem/kubeapi_test.go]: testing this

File: subsystem/kubeapi_test.go (hunk 2 above)
Comment (filanov) [reply subsystem/kubeapi_test.go:767]: But once people will add more tests will become a mess no? why not check the deployment type in the aftereach and react according to it? 
All the tests need to clear the DB so why not write it once?

Comment (filanov) [reply subsystem/kubeapi_test.go:767]: maybe something like that https://play.golang.org/p/7_zCj4Lqawr

Comment (nmagnezi) [reply subsystem/kubeapi_test.go:767]: Per discussion @filanov and I just had, I'll be doing this in a separate PR. I Opened [MGMT-8299](https://issues.redhat.com/browse/MGMT-8299) to track this.

File: subsystem/kubeapi_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (filanov) [reply subsystem/kubeapi_test.go]: https://onsi.github.io/ginkgo/ take a look at `Logging Output` section

[the remark this answers is not in the document]
Comment (nmagnezi) [reply subsystem/kubeapi_test.go]: I experimented with GinkgoWriter [here](https://github.com/openshift/assisted-service/pull/2952/files#diff-f2d3cebc87fae81285493762229b8e7ecac6ab614ea0d82950af55f303dcc3d5R568-R647), results [look good ](
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_assisted-service/2952/pull-ci-openshift-assisted-service-master-subsystem-kubeapi-aws/1462525072416706560)(open stdout)

**Review by osherdp — APPROVED** [review]

File: subsystem/kubeapi_test.go [hunk 4]
```diff
@@ -564,6 +565,86 @@ func randomNameSuffix() string {
...
+			clusterImageSetList   hivev1.ClusterImageSetList
+			clusterDeploymentList hivev1.ClusterDeploymentList
+		)
+
+		multiErr = multierror.Append(multiErr, client.List(ctx, &agentList, k8sclient.InNamespace(Options.Namespace)))
+		aList, err := json.Marshal(agentList)
+		if err != nil {
+			multiErr = multierror.Append(multiErr, err)
+		} else {
+			_, err = GinkgoWriter.Write([]byte(fmt.Sprintf("failed test '%s' created the agents: %s\n", GinkgoT().Name(), aList)))
```
Comment (filanov) [inline subsystem/kubeapi_test.go:586]: Wrap it in a function that will get a string or format
I think that it can be treated as a log function

**Review by osherdp — APPROVED** [review]

**Review by osherdp — APPROVED** [review]

File: subsystem/kubeapi_test.go (hunk 4 above)
Comment (nmagnezi) [reply subsystem/kubeapi_test.go]: Done

**Comment by nmagnezi** [comment]
/unhold
I was able to improve this a bit more.
It now prints the created resources with indentation, which is more friendly to the reader. Examples (click open stdout):
* [subsystem-kubeapi-aws](
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_assisted-service/2952/pull-ci-openshift-assisted-service-master-subsystem-kubeapi-aws/1462823417949458432)
* [subsystem-aws](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_assisted-service/2948/pull-ci-openshift-assisted-service-master-subsystem-aws/1462802027338797056)

**Comment by nmagnezi** [comment]
> /retest-required

We seem to have a flaky test here. I opened [MGMT-8307](https://issues.redhat.com/browse/MGMT-8307) to track it.

**Pull request merged**